### ISSUE TITLE
Extended 'file' custom widget doesn't get the uploaded file

### ISSUE THREAD

**Issue by MelMacaluso** [netlify/netlify-cms#2955]
**Describe the bug**
I am extending the `file` widget through `getWidget('file')` to achieve some custom validation of an uploaded .json. After some guessing/research I found out just class components are supported: 

```js
import CMS from 'netlify-cms-app'
import React, { PureComponent } from 'react'

const FileControl = CMS.getWidget('file')

class JSONControl extends PureComponent {
  isValid = () => ({ error: { message: 'Your error message.' } })

  render() {
    return <FileControl.control {...this.props} />
  }
}


export default JSONControl
```

**To Reproduce**
When I try to upload a file after clicking "choose file" and selecting it from the media library (uploads successfully) it doesn't get assigned as value to the field and the field still shows "choose file" as if nothing happened. 

**Expected behavior**
My aim here is to upload such file and on "publish" perform the `isValid` validation that I will implement. Therefore I need to have access to the file. 

**Applicable Versions:**
netlify-cms-core 2.12.1
netlify-cms-app 2.9.6

**CMS configuration**
The relevant CMS config for that file is: 
```
            {
              label: 'JSON',
              name: 'JSON',
              widget: 'JSON',
            },
```

**Comment by barthc** [comment]
@MelMacaluso  try something like so:

```js
import CMS from 'netlify-cms'

const File = CMS.getWidget('file')

class JSONControl extends File.control {
  isValid = () => {
    // Do internal validation
    return { error: { message: 'Your error message.' } };
  };
}


export default JSONControl
```

**Comment by sleepingkiwi** [comment]
I was struggling with this today as well.

@barthc has a good solution if you don't need to wrap the original widget.

I needed to make this work by wrapping the widget with code like in @MelMacaluso original example. For this you just need to provide your own `shouldComponentUpdate` method:

```js
class JSONControl extends PureComponent {
  isValid = () => ({ error: { message: 'Your error message.' } })
  
  shouldComponentUpdate() {
    // this would work
    return true;

    // so would this
    /*
    return (
      this.props.value !== nextProps.value
      || this.props.classNameWrapper !== nextProps.classNameWrapper
      || this.props.hasActiveStyle !== nextProps.hasActiveStyle
      || this.props.mediaPaths !== nextProps.mediaPaths
    );
    */
  }

  render() {
    return <FileControl.control {...this.props} />
  }
}
```

*In case anyone cares, and to help people searching for keywords to resolve similar issues:*

The `image` and `file` widgets both implement [`withFileControl.js`](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-widget-file/src/withFileControl.js) which relies on the value of the `mediaPaths` prop changing to render any changes. 

Widgets that are registered with `CMS.registerWidget()` and *don't* provide their own `shouldComponentUpdate` method get given a default, I believe from [`Widget.js`](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-core/src/components/Editor/EditorControlPane/Widget.js#L61). The default does not update when the value of `mediaPaths` changes so your wrapped widget never gets the update.

Same solution referenced in #2700

**Comment by MelMacaluso** [comment]
Hi @barthc thanks a lot, I tried to make your snippet work to no avail. Reason could be because I am using `netlify-cms-app`?

@sleepingkiwi Thanks, it indeed worked but defies my realm of understanding on why the `shouldComponentUpdate()` is needed...

I finally see the file "uploaded" in the backend (showing the filepath such as `/assets/file.json`). 

I naively thought then I could get it with a fetch request from the backend using that path and do my validation, it indeed doesn't work. Any clues on how to achieve that? Apologies about the off toppic as well!

**Comment by sleepingkiwi** [comment]
The simple answer is to use `this.props.getAsset()` with the caveat that unless you're using a media library integration like [Cloudinary](https://www.netlifycms.org/docs/cloudinary/) you might still run into problems with long deploy times.

If you need to load the file contents you probably specifically need `this.props.getAsset(this.props.value).toString()` which returns either the full filepath or *if the media was uploaded in the same session* (i.e. you just uploaded it) a locally cached `blob` which you can use to load the full file contents.

This means you should be able to validate the file immediately after uploading.

The caveat is that, unless you're using a media library integration, if you upload a file and then refresh the page you will no longer be able to validate (or preview) it until after your deploy has finished because `props.getAsset()` will start returning the full file path where it will eventually live when the deploy has built. In a lot of cases that's fine but it can lead to some weird situations if the person using the CMS isn't prepared for it.

I haven't found a good way around this without editing the core files. There is a file path to the asset on github (or whatever backend you're using) stored in the redux store but I don't think we can access that from a custom widget.

**Comment by erezrokah** [comment]
@sleepingkiwi FYI, I'm actually refactoring the way the CMS is handling assets as a part of https://github.com/netlify/netlify-cms/issues/1472.

There are numerous bugs related to assets not shown in the CMS UI as they have not been published yet to the site:
https://github.com/netlify/netlify-cms/issues/2953
https://github.com/netlify/netlify-cms/issues/1750
https://github.com/netlify/netlify-cms/issues/2332
https://github.com/netlify/netlify-cms/issues/2516

In short the `getAsset` will try to retrieve the asset from the UI state and fallback to get it from the repo. That meant changing `getAsset` to return a promise that resolves with the asset url instead of just a string.

**Comment by sleepingkiwi** [comment]
Fantastic, thanks for your work on that @erezrokah I think that's exactly how I'd want getAsset to work for our use cases. I'm excited to see that land!

**Comment by MelMacaluso** [comment]
@sleepingkiwi  and @erezrokah  thanks for the pointers.

Previously to posting this issue I tried the `getAsset()` method as I saw it in the props (where is all of this documented? I am doing so much detecting work ahah) but after I upload the .json and try to submit I get `undefined` or `null` (can't recall) from `getAsset()` and apparently the asset is correctly uploaded...

While if I wait until it has been pushed to the repo it returns the file itself but that prevents me to do on the fly validation...

In terms of what I am using to upload files, I didn't setup anything myself just used the [gatsby-netlify-cms starter](https://github.com/netlify-templates/gatsby-starter-netlify-cms).

Any top level suggestions would be greatly appreciated. Thanks in advance for your hard work 🙏

**Comment by barthc** [comment]
@MelMacaluso the file would have already been uploaded to the repo if you are using the widget `isValid` function for validation. This widget validation is just to assign the correct front-matter value.

If you are thinking about validating the file before upload, that would be an issue for the media library. I think there is already an issue for validating image files.

**Comment by MelMacaluso** [comment]
> @MelMacaluso the file would have already been uploaded to the repo if you are using the widget `isValid` function for validation. This widget validation is just to assign the correct front-matter value.
> 
> If you are thinking about validating the file before upload, that would be an issue for the media library. I think there is already an issue for validating image files.

Thanks. So what you are saying is that I should perform the validation when it is uploaded in the actual library (the .json file being the subject). 

I will try to find somewhere where/how that can be done (if you happen to have a link to the docs it would be immensely helpful). Thanks!

**Comment by barthc** [comment]
> I will try to find somewhere where/how that can be done (if you happen to have a link to the docs it would be immensely helpful). Thanks!

Unfortunately, there is no way to do that at the moment, sorry. Here is the issue for image validation #1322

**Comment by MelMacaluso** [comment]
> > I will try to find somewhere where/how that can be done (if you happen to have a link to the docs it would be immensely helpful). Thanks!
> 
> Unfortunately, there is no way to do that at the moment, sorry. Here is the issue for image validation #1322

Not to worry, thanks as lot mate!
